diff --git a/hopper/flash_api.cpp b/hopper/flash_api.cpp
index 43f3387..56b7a9e 100644
--- a/hopper/flash_api.cpp
+++ b/hopper/flash_api.cpp
@@ -2,39 +2,19 @@
  * Copyright (c) 2024, Jay Shah, Ganesh Bikshandi, Ying Zhang, Vijay Thakkar, Pradeep Ramani, Tri Dao.
  ******************************************************************************/
 
-#include <Python.h>
-#include <torch/nn/functional/padding.h>
 #include <ATen/cuda/CUDAContextLight.h>
 #include <c10/cuda/CUDAGuard.h>
+#include <torch/nn/functional/padding.h>
 
 #include <cutlass/numeric_types.h>
 
+#include "flash_api.h"
 #include "flash.h"
 #include "static_switch.h"
 #include "tile_size.h"
 #include "heuristics.h"
 #include "cuda_check.h"
 
-
-extern "C" {
-/* Creates a dummy empty _C module that can be imported from Python.
-    The import from Python will load the .so consisting of this file
-    in this extension, so that the TORCH_LIBRARY static initializers
-    below are run. */
-PyObject* PyInit__C(void)
-{
-    static struct PyModuleDef module_def = {
-        PyModuleDef_HEAD_INIT,
-        "_C",   /* name of module */
-        NULL,   /* module documentation, may be NULL */
-        -1,     /* size of per-interpreter state of the module,
-                    or -1 if the module keeps state in global variables. */
-        NULL,   /* methods */
-    };
-    return PyModule_Create(&module_def);
-}
-}
-
 #define CHECK_DEVICE(x) TORCH_CHECK(x.is_cuda(), #x " must be on CUDA")
 #define CHECK_SHAPE(x, ...) TORCH_CHECK(x.sizes() == torch::IntArrayRef({__VA_ARGS__}), #x " must have shape (" #__VA_ARGS__ ")")
 #define CHECK_CONTIGUOUS(x) TORCH_CHECK(x.is_contiguous(), #x " must be contiguous")
diff --git a/hopper/flash_api_stable.cpp b/hopper/flash_api_stable.cpp
index 9759af8..7a88468 100644
--- a/hopper/flash_api_stable.cpp
+++ b/hopper/flash_api_stable.cpp
@@ -2,8 +2,6 @@
  * Copyright (c) 2024, Jay Shah, Ganesh Bikshandi, Ying Zhang, Vijay Thakkar, Pradeep Ramani, Tri Dao.
  ******************************************************************************/
 
-#include <Python.h>
-
 #include <cutlass/numeric_types.h>
 
 #include "flash.h"
@@ -78,26 +76,6 @@ cudaDeviceProp* get_device_prop() {
 }
 } // anonymous namespace
 
-
-extern "C" {
-/* Creates a dummy empty _C module that can be imported from Python.
-    The import from Python will load the .so consisting of this file
-    in this extension, so that the STABLE_TORCH_LIBRARY static initializers
-    below are run. */
-PyObject* PyInit__C(void)
-{
-    static struct PyModuleDef module_def = {
-        PyModuleDef_HEAD_INIT,
-        "_C",   /* name of module */
-        NULL,   /* module documentation, may be NULL */
-        -1,     /* size of per-interpreter state of the module,
-                    or -1 if the module keeps state in global variables. */
-        NULL,   /* methods */
-    };
-    return PyModule_Create(&module_def);
-}
-}
-
 #define CHECK_DEVICE(x) STD_TORCH_CHECK(x.is_cuda(), #x " must be on CUDA")
 #define CHECK_SHAPE(x, ...) \
     do { \
diff --git a/hopper/flash_api.h b/hopper/flash_api.h
new file mode 100644
index 0000000..48da7b7
--- /dev/null
+++ b/hopper/flash_api.h
@@ -0,0 +1,43 @@
+#pragma once
+
+#include <ATen/ATen.h>
+#include <cstdint>
+
+#include <optional>
+#include <tuple>
+
+std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor> mha_fwd(
+        at::Tensor q,
+        at::Tensor k,
+        at::Tensor v,
+        std::optional<at::Tensor> k_new_,
+        std::optional<at::Tensor> v_new_,
+        std::optional<at::Tensor> q_v_,
+        std::optional<at::Tensor> out_,
+        std::optional<at::Tensor> cumsum_seqlens_q_,
+        std::optional<at::Tensor> cumsum_seqlens_k_,
+        std::optional<at::Tensor> cumsum_seqlens_k_new_,
+        std::optional<at::Tensor> seqused_q_,
+        std::optional<at::Tensor> seqused_k_,
+        std::optional<int64_t> max_seqlen_q_,
+        std::optional<int64_t> max_seqlen_k_,
+        std::optional<at::Tensor> page_table_,
+        std::optional<at::Tensor> kv_batch_idx_,
+        std::optional<at::Tensor> leftpad_k_,
+        std::optional<at::Tensor> rotary_cos_,
+        std::optional<at::Tensor> rotary_sin_,
+        std::optional<at::Tensor> seqlens_rotary_,
+        std::optional<at::Tensor> q_descale_,
+        std::optional<at::Tensor> k_descale_,
+        std::optional<at::Tensor> v_descale_,
+        std::optional<double> softmax_scale_,
+        bool is_causal,
+        int64_t window_size_left,
+        int64_t window_size_right,
+        int64_t attention_chunk,
+        double softcap,
+        bool is_rotary_interleaved,
+        std::optional<at::Tensor> scheduler_metadata_,
+        int64_t num_splits,
+        std::optional<bool> pack_gqa_,
+        int64_t sm_margin);
